articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+[To discuss: Should there be a callback for when configuration is changed, e.g. altered parameters, new topics, etc.?]
+
+While in this state, the node will not receive any execution time to read topics, perform processing of data, respond to functional service requests, etc.
+
+In the inactive state, any data that arrives on topics will not be read. Whether or not data is queued will depend on the QoS parameters of the topics involved. [In other words, to queue data or not is up to each topic.]
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
```

Comment:
Just to make things clearer: as I thought I said twice, this is a *system design* matter. What I really would wish is that this paragraph gets removed from a document whose concern is node lifecycle design.

Now, that's an interesting discussion (even though it's one I had so many times already ...)

> By what definition ?

Mine, admittedly ;-) This is indeed a problem of point of view. The node's "recovered error" is a system's "nominal execution". A node's exception can be a system's recovered error.

> If the node has no error recovery logic in it then you're saying that all error recovery logic must either be in the supervision code (which breaks the encapsulation of the node)

Depends on what is required to recover. The strategy of 98% of all drivers (well, actually, let's make that 100% if we ignore rounding errors) is "close and reopen". That's something that can *definitely* be embedded in a supervision layer. What the supervision layer can more easily do is escalate (turn power off/on after a few tries if the system has this capability) and *more importantly* it can make sure that the rest of the system does not need this device while it's being tried.

If you actually have non-hypothetical recovery mechanisms that involve complex cryptic dances, then they probably should be in the driver. But I have yet to see anything like that in the wild. Moreover, I would integrate that as an optional feature that the supervision would have to turn on, simply because it means that the device is not available and probably won't be in the foreseeable future (and will take longer than expected to initialize). If you have a DVL in this state with your AUV near a fragile structure, let's say that you will definitely want to mitigate the unavailability of the DVL.

My experience is that:
 - RTT has a runtime error state, which I never saw used in any form of meaningful way
 - I never saw a complex device recovery mechanism in the wild (that is, apart from device reset which is done in Rock during the transition to exception). A lot of very robust devices though, of all types. Most drivers that had spurious device errors were due to a completely nonsensical handling of I/O (i.e. random errors because of non-existent buffering before packet parsing) and their recovery strategy was "reopen the I/O and start from scratch".
 - at higher level, errors are either due to a bad interaction between nodes, i.e. their recovery is a system concern or is made of things that can't be meaningfully recovered by the node (e.g. a KF gone mad).
 - the most meaningful device recovery mechanism (as in "recover from non-functioning device") I saw is one that powers the device off and on and reconfigures it. System-level as it involves synchronization between different devices (the node that handles the power distribution).


> @doudou: you can program your own drivers however you like, but let's not dictate those kinds of low-level details so early in ROS 2 system design.

My point, exactly. Which is why I react on it on a document that defines the *node lifecycle* design (not even the *system* design). I.e. would prefer that this gets removed.

> Many ROS 1 device drivers support dynamic_reconfigure, and that is very convenient for users, who can fine-tune device parameters while continuously observing the results.

I fail to see how that's relevant. If the change of parameter requires a complete reset of the connection to the device, this is done in Rock by externally cycling through the state machine and change the parameter as static. Simply because, again, it makes the node's implementation that much simpler (reduces the number of codepaths) *and* informs the rest of the system of the effect of changing parameter (i.e. possibly very costly). If there is actually a device functionality for changing the parameter "live" (e.g. a camera brightness), then this is done dynamically. Both of them can be handled transparently from a user point of view.